python/packages/ai/teams/ai/augmentations/tools_augmentation.py

```diff
@@ -0,0 +1,115 @@
+"""
+Copyright (c) Microsoft Corporation. All rights reserved.
+Licensed under the MIT License.
+"""
+
+from __future__ import annotations
+
+import json
+from typing import List, Optional, Union, cast
+
+from botbuilder.core import TurnContext
+
+from ...state import MemoryBase
+from ..models.chat_completion_action import ChatCompletionAction
+from ..models.prompt_response import PromptResponse
+from ..planners.plan import (
+    Plan,
+    PredictedCommand,
+    PredictedDoCommand,
+    PredictedSayCommand,
+)
+from ..prompts.message import ActionCall, Message
+from ..prompts.sections.prompt_section import PromptSection
+from ..tokenizers.tokenizer import Tokenizer
+from ..validators.validation import Validation
+from .augmentation import Augmentation
+from .tools_constants import ACTIONS_HISTORY
+
+
+class ToolsAugmentation(Augmentation[Union[str, List[ActionCall]]]):
+    """
+    A server-side 'tools' augmentation.
+    """
+
+    _actions: Optional[List[ChatCompletionAction]]
+
+    def __init__(self, actions: Optional[List[ChatCompletionAction]]) -> None:
+        self._actions = actions
+
+    def create_prompt_section(self) -> Optional[PromptSection]:
+        """
+        Creates an optional prompt section for the augmentation.
+        """
+        return None
+
+    async def validate_response(
+        self,
+        context: TurnContext,
+        memory: MemoryBase,
+        tokenizer: Tokenizer,
+        response: PromptResponse[Union[str, List[ActionCall]]],
+        remaining_attempts: int,
+    ) -> Validation:
+        """
+        Validates a response to a prompt.
+
+        Args:
+            context (TurnContext): Context for the current turn of conversation.
+            memory (MemoryBase): Interface for accessing state variables.
+            tokenizer (Tokenizer): Tokenizer to use for encoding/decoding text.
+            response (PromptResponse[Union[str, List[ActionCall]]]):
+                Response to validate.
+            remaining_attempts (int): Nubmer of remaining attempts to validate the response.
+
+        Returns:
+            Validation: A 'Validation' object.
+        """
+        if response.message and response.message.action_calls and memory.has(ACTIONS_HISTORY):
+            tool_calls = response.message.action_calls
+            tools = self._actions
+
+            if tools and len(tool_calls) > 0:
+                return Validation(
+                    valid=True,
+                    value=tool_calls,
+                )
+
+        memory.set(ACTIONS_HISTORY, [])
```

Comment:
whenever the LLM responds you clear the conversation history?

Reply:
cleared if the LLM responds with no tools (for sanity purposes to ensure a clean slate for subsequent calls)